compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -0,0 +1,368 @@
+                                               applyOp.getMapOperands())
+                .getResult();
+      } else if (auto constantOffset = getConstantIntValue(l3SourceOffsetVal)) {
+        int64_t newOffset = *constantOffset + offsetToAdd;
+        newL3AsSourceOffset = rewriter.getIndexAttr(newOffset);
+      } else {
+        return failure();
+      }
+    }
+  }
+  return newL3AsSourceOffset;
+}
+
+// Given a vector of L2->L1 Dma ops, split them via the following process :-
+// 1. Infer splitting dimension of L2 as a function of all L2->L1 Dma ops.
+// 2. Fetch and ensure a unique L3->L2 Dma op.
+// 3. For the split dimension inferred set offset = 0 and size as 1 for L2 and
+// L3.
+// 4. Now traverse each L2->L1 Dma op and perform the following :-
+//    a) Create a new L2 AllocOp based on the updated size (step 3 above) and
+//    create
+//       a logicalobjectfifo using the same.
+//    b) Split L3->L2 Dma op.
+//    c) SPlit L2->L1 Dma op.
+LogicalResult splitLogicalObjectFifos(
+    IRRewriter &rewriter, SmallVector<AMDAIE::DmaCpyNdOp> &l2ToL1DmaOps,
+    MLIRContext *context) {
+  if (l2ToL1DmaOps.size() == 0) return success();
+
+  SmallVector<OpFoldResult> baseSourceOffsets =
+      l2ToL1DmaOps[0].getSourceMixedOffsets();
+  LogicalObjectFifoFromMemrefOp sourceObjectFifo =
+      l2ToL1DmaOps[0].getSourceObjectFifo();
+  auto sourceAllocOp =
+      sourceObjectFifo.getMemref().getDefiningOp<memref::AllocOp>();
+  if (!sourceAllocOp) {
+    sourceObjectFifo->emitOpError()
+        << "expected alloc op as the defining op of source "
+           "logicalobjectfifo.from_memref";
+    return failure();
+  }
+  // We will now capture those dimensions where L2 memory was split. The way we
+  // do this is by checking all L2->L1 DmaOps' source offset and marking those
+  // dimensions which are not equal to at least one of the source offsets.
+  DenseSet<size_t> splitDimsSetForL2;
+  SmallVector<size_t> splitDimsForL2;
+  for (unsigned i = 1, n = l2ToL1DmaOps.size(); i < n; i++) {
+    if (l2ToL1DmaOps[i].getSourceObjectFifo() != sourceObjectFifo) {
+      l2ToL1DmaOps[i]->emitRemark() << "has different source objectfifo";
+      sourceObjectFifo->emitRemark() << "is the expected source objectfifo";
+      return failure();
+    }
+    SmallVector<OpFoldResult> sourceOffsets =
+        l2ToL1DmaOps[i].getSourceMixedOffsets();
+    for (unsigned j = 0, m = baseSourceOffsets.size(); j < m; j++) {
+      if (baseSourceOffsets[j] != sourceOffsets[j] &&
+          !splitDimsSetForL2.contains(j)) {
+        splitDimsForL2.push_back(j);
+        splitDimsSetForL2.insert(j);
+      }
+    }
+  }
+  std::sort(splitDimsForL2.begin(), splitDimsForL2.end());
+
+  if (failed(checkIsRangeFromZero(splitDimsForL2))) {
+    l2ToL1DmaOps[0]->emitRemark()
+        << "cannot split L2 logicalobjectfifo because of non-contiguous split "
+           "dimensions inferred";
+    return failure();
```

Comment:
Why use `emitRemark`? Imo, this should be an error:

```suggestion
    return l2ToL1DmaOps[0]->emitOpError()
        << "cannot split L2 logicalobjectfifo because of non-contiguous split "
           "dimensions inferred";
```